db/src/lib.rs

```diff
@@ -35,6 +32,34 @@ pub async fn create_user(pool: &PgPool, user_name: &str) -> color_eyre::Result<U
     Ok(result.user_id)
 }
 
+pub async fn add_url(pool: &PgPool, url: &str, allow_existing: &bool) -> color_eyre::Result<String> {
```

Comment:
Your implementation is great! And exactly what I would have expected!

Butt looking at with 20 years of experience I think we can optimize it! And honestly it probably doesn't _need_ to be optimized now, and I'm going to break my rule of not optimizing Rust until we know we need it. BUTT I want to chat about this and share 😆 !

Right now this requires two SQL calls to add the URL, one to check if it exists and one to actually add it if needed. And I THINK we can drop this down to a single SQL call in all cases 🎉 

But it is going to be a bit weird/different so please ask questions if it doesn't make sense!

The basic idea is that we can do what is called an `upsert` to either insert or update the record in a single SQL call, and then after that we can tell if the record was just created or not! This 'upsert' will also rely on the unique index on `(user_id, url)` that I mentioned in another comment!

The idea would be something like this code!

```rust
let new_page_id = Uuid::gen_v4();

let upsert_result = sqlx::query_as!(
Pages,
"INSERT INTO Page (page_id, user_id, url) VALUES ($1, $2, $3) RETURNING * ON CONFLICT DO NOTHING",
new_page_id,
user_id,
url)
.fetch_one(pool)
.execute();

let exists = upsert_result.page_id != new_page_id;
```

The basic idea is that we can create a new `page_id` in Rust (instead of letting PG make one for us) and then try to insert the URL with that `page_id` and with `ON CONFLICT DO NOTHING` at the end of the insert. This tells Postgres that if it gets a conflict on a unique index we want it to 'DO NOTHING' or not update any columns.

This means that if we inserted a new Page it will have the `page_id` we specified. But if this User already added this url it will have a _DIFFERENT_ if since the record already existed.

We can check this after the insert to know if it was a new record or not, and error out if the allow_existing flag wasn't passed in!

Reply:
It was journey but I enjoyed it